packages/core/src/stateUtils.ts

```diff
@@ -1508,87 +1485,179 @@ export function resolveTransition<
+        if (sendAction.to === SpecialTargets.Internal) {
+          raisedActions.push(sendAction as RaiseActionObject<TEvent>);
+        } else {
+          resActions.push(sendAction);
+        }
+        break;
+      case actionTypes.log:
+        resActions.push(
+          resolveLog(
+            actionObject as LogAction<TContext, TEvent>,
+            context,
+            _event
+          )
+        );
+        break;
+      case actionTypes.pure:
+        resActions.push(
+          ...toArray(
+            (actionObject as PureAction<TContext, TEvent>).get(
+              context,
+              _event.data
+            )
+          )
+        );
+        break;
+      case actionTypes.assign:
+        context = updateContext(
+          context,
+          _event,
+          [actionObject as AssignAction<TContext, TEvent>],
+          currentState
+        );
+        resActions.push(actionObject);
+        break;
+      default:
+        resActions.push(
+          toActionObject(actionObject, machine.machine.options.actions)
+        );
+        break;
     }
+  });
 
-    while (raisedEvents.length) {
-      const raisedEvent = raisedEvents.shift()!;
-      maybeNextState = resolveRaisedTransition(
-        machine,
-        maybeNextState,
-        raisedEvent._event,
-        _event
-      );
-    }
-  }
+  return {
+    actions: resActions,
+    raised: raisedActions,
+    context
+  };
+}
+
+export function macrostep<TContext, TEvent extends EventObject>(
+  state: State<TContext, TEvent, any, Typestate<TContext>>,
+  event: Event<TEvent> | SCXML.Event<TEvent> | null,
+  machine: MachineNode<TContext, any, TEvent, any>
+): State<TContext, TEvent> {
+  // Assume the state is at rest (no raised events)
+  // Determine the next state based on the next microstep
+  const nextState = event === null ? state : machine.microstep(state, event);
+
+  const { _event, _internalQueue } = nextState;
+  let maybeNextState = nextState;
 
-  // Detect if state changed
-  const changed =
-    maybeNextState.changed ||
-    (history
-      ? !!maybeNextState.actions.length ||
-        !!assignActions.length ||
-        typeof history.value !== typeof maybeNextState.value ||
-        !stateValuesEqual(maybeNextState.value, history.value)
-      : undefined);
+  while (_internalQueue.length && !maybeNextState.done) {
+    const raisedEvent = _internalQueue.shift()!;
+    const currentActions = maybeNextState.actions;
 
-  maybeNextState.changed = changed;
+    maybeNextState = machine.microstep(
+      maybeNextState,
+      raisedEvent as SCXML.Event<TEvent>
+    );
 
-  // TODO: remove children if they are stopped
-  maybeNextState.children = children;
+    _internalQueue.unshift(...maybeNextState._internalQueue);
 
-  // Preserve original history after raised events
-  maybeNextState.history = history;
-  maybeNextState.historyValue = resolved.historyValue;
+    // Save original event to state
+    if (raisedEvent.type === NULL_EVENT) {
```

Comment:
this has a weird effect that quite different event gets saved on the `maybeNextState`, as in this case you restore the very first event which has caused a macrostep, but in other cases you let it be updated with the most recent raised event - and null "events" and raised events might occur in any order and multiple times, so this leads to very inconsistent results

probably you should always save the most recent raised event here (apart from null events)

Reply:
Fixed - it will now use the most recently raised event to overwrite `NULL_EVENT`.